### PR TITLE
New SourceHitCountsCache (FE-1277)

### PR DESCRIPTION
**Pull request by hbenl**

### CODE REVIEW

**Review by bvaughn — CHANGES_REQUESTED** [review]
Can you hold off on this PR for now?

It looks like it's going to conflict with #8884 which I've been working on for a few days.

**Comment by bvaughn** [comment]
The "suspense" package has been added now (#8909) and I've migrated a few example caches to it (#8910, #8936, and WIP #8940). Can we add this new cache using it? 😄

**Comment by jasonLaster** [comment]
@hbenl i think it's fine to show the hit counts for the currently loaded window. It's cool that the backend `getSymbols` call could contain all hit counts, but to be consistent with the rest of devtools we really should only show the hit counts for the current window

**Comment by bvaughn** [comment]
> That's the plan. However it seems that the interval cache currently doesn't expose what I need to create hooks for it (i.e. I don't see a way to get cached values and to subscribe to changes) but I need that.

I started adding that last night, FWIW. Let me finish (this morning) first and then pick this back up.

**Comment by hbenl** [comment]
> i think it's fine to show the hit counts for the currently loaded window.

I think there's a misunderstanding: I'm not talking about the loaded range of the recording but about the part of the source visible in the source panel. The Outline panel will show 0 hits for functions that are not currently visible in the source panel.
I'm going to change the UI to show no hitcount badge (instead of a hitcount badge with content "0") for functions that are not visible in the source panel. That was the part that I found confusing.

**Comment by bvaughn** [comment]
>> Can we add this new cache using it?
>
> That's the plan. However it seems that the interval cache currently doesn't expose what I need to create hooks for it (i.e. I don't see a way to get cached values and to subscribe to changes) but I need that.

Released 0.0.26 with new [`useIntervalCacheStatus` hook](https://suspense.vercel.app/useIntervalCacheStatus) which should unblock this. I'll post a PR shortly that upgrades us.

**Comment by hbenl** [comment]
about test failures:
- the replay bot comment has not been updated after the last screenshot test run: the tests actually all passed
- the changed screenshots show one logpoint has moved slightly - I guess that the backend gave me a slightly different `time` value when I recreated some protocol fixtures

File: src/ui/components/Timeline/PreviewMarkers.tsx
Comment (bvaughn) [inline src/ui/components/Timeline/PreviewMarkers.tsx:37]: Can we use our binary search util here to make this faster?

File: packages/replay-next/src/suspense/SourceHitCountsCache.ts
Comment (bvaughn) [inline packages/replay-next/src/suspense/SourceHitCountsCache.ts:91]: If I'm understanding it correctly– the only reason it exists is to let hit counts be loaded _without_ using Suspense? (else we'd be using `sourceHitCountsCache.read` instead of `sourceHitCountsCache.readAsync`)

I don't think we need it. Should be able to just use `sourceHitCountsCache` directly.

File: packages/replay-next/components/sources/SourceList.tsx
Comment (bvaughn) [inline packages/replay-next/components/sources/SourceList.tsx:103]: Why do we need `useGetSourceHitCounts`? I think we can just use `sourceHitCountsCache.read` directly here.

The only place `SourceList` is used is inside of `<Source><Suspense><SourceLoader>`

File: packages/replay-next/src/suspense/SourceHitCountsCache.ts
Comment (hbenl) [reply packages/replay-next/src/suspense/SourceHitCountsCache.ts:91]: > If I'm understanding it correctly– the only reason it exists is to let hit counts be loaded without using Suspense? 

Correct.

> I don't think we need it. Should be able to just use sourceHitCountsCache directly.

Yes, that would be possible, but this is a performance optimization we consciously added in #8695.

Comment (bvaughn) [reply packages/replay-next/src/suspense/SourceHitCountsCache.ts:91]: I see. So the goal here is not so much to bypass Suspense– but to complete the initial render _without suspending_ if data hasn't let been loaded. The `useImperativeCacheValue` hook (in "suspense") does this, although that's not its main purpose– and it doesn't work with interval caches.

Maybe I should just add a `useIntervalCacheValue` hook to the "suspense" package too.

Comment (hbenl) [reply packages/replay-next/src/suspense/SourceHitCountsCache.ts:91]: > Maybe I should just add a useIntervalCacheValue hook to the "suspense" package too.

Yes, I think that would make sense.

Comment (bvaughn) [reply packages/replay-next/src/suspense/SourceHitCountsCache.ts:91]: Oh it.

Comment (bvaughn) [reply packages/replay-next/src/suspense/SourceHitCountsCache.ts:91]: PR #8968 updates `suspense@0.0.28` which adds a new hook, `useImperativeIntervalCacheValues`:
https://suspense.vercel.app/useImperativeIntervalCacheValues

Should be usable here 😄

Comment (hbenl) [reply packages/replay-next/src/suspense/SourceHitCountsCache.ts:91]: I'll rebase when that PR is fixed and merged.

File: packages/replay-next/playwright/tests/object-inspector.ts
Comment (markerikson) [inline packages/replay-next/playwright/tests/object-inspector.ts:182]: :hand: Need to remove this `.only`

Comment (hbenl) [reply packages/replay-next/playwright/tests/object-inspector.ts:182]: D'oh!

File: packages/replay-next/src/suspense/SourceHitCountsCache.ts
Comment (bvaughn) [reply packages/replay-next/src/suspense/SourceHitCountsCache.ts:91]: Should be ready now!

**Review by markerikson — APPROVED** [review]

**Pull request merged**
